### ISSUE TITLE
Too many lookup parameters in populated model

### ISSUE THREAD

**Issue by peterparkero** [clab/dynet#1221]
TLDR: Upon loading a model from a file that I have just built, the programme throws the following error:

> RuntimeError: Too many lookup parameters in populated model at /_21

Specifically, the follow code was adapted from https://github.com/jcyk/greedyCWS
(The original code works under Python 2 and Dynet 1.0, however, I have adapted part of the code to make it work under Python 3 and Dynet 2.0)

Character embeddings is first loaded, where it will then be used to initialise the CWS object.  This CWS object is a dynet model object with the specified structured defined under initParams.  Suppose I specify the saved model file under load_params, the CWS object will attempt to populate the model based on the previously saved files containing the model using CWS.save(filename) where self.model.save(filename) will be called subsequently.  This is where the runtime error comes in.  I have traced this error to be originated from dynet/io.cc, however, I have no clue why this error is prompted, or how I can adapt this code to make the programme work.

    Cemb, character_idx_map = initCemb(char_dims,train_file,pre_trained)
    cws = CWS(Cemb,character_idx_map,options)

    if load_params is not None:
        cws.load(load_params)
        test(cws, dev_file, 'result')

where CWS is defined as:

    class CWS (object):
        def __init__(self,Cemb,character_idx_map,options):
            model = dy.Model()
            self.trainer = dy.MomentumSGDTrainer(model,options['lr'],options['momentum'])
            self.params = self.initParams(model,Cemb,options)
            self.options = options
            self.model = model
            self.character_idx_map = character_idx_map
            self.known_words = None

        def load(self,filename):
            self.model.populate(filename)

        def save(self,filename):
            self.model.save(filename)

        def use_word_embed(self,known_words):
            self.known_words = known_words
            self.params['word_embed'] = self.model.add_lookup_parameters((len(known_words),self.options['word_dims']))

        def initParams(self,model,Cemb,options):
            # initialize the model parameters  
            params = dict()
            params['embed'] = model.add_lookup_parameters(Cemb.shape)
            for row_num,vec in enumerate(Cemb):
                params['embed'].init_row(row_num, vec)
            params['lstm'] = dy.LSTMBuilder(1,options['word_dims'],options['nhiddens'],model)
        
            params['reset_gate_W'] = []
            params['reset_gate_b'] = []
            params['com_W'] = []
            params['com_b'] = []

            params['word_score_U'] = model.add_parameters(options['word_dims'])
            params['predict_W'] = model.add_parameters((options['word_dims'],options['nhiddens']))
            params['predict_b'] = model.add_parameters(options['word_dims'])
            for wlen in range(1,options['max_word_len']+1):
                params['reset_gate_W'].append(model.add_parameters((wlen*options['char_dims'],wlen*options['char_dims'])))
                params['reset_gate_b'].append(model.add_parameters(wlen*options['char_dims']))
                params['com_W'].append(model.add_parameters((options['word_dims'],wlen*options['char_dims'])))
                params['com_b'].append(model.add_parameters(options['word_dims']))
            params['<BoS>'] = model.add_parameters(options['word_dims'])
            return params

The headers of the model file are as follows: 

> #Parameter# /vanilla-lstm-builder/_0 {200,50} 320002 FULL_GRAD
> #Parameter# /vanilla-lstm-builder/_1 {200,50} 320002 FULL_GRAD
> #Parameter# /vanilla-lstm-builder/_2 {200} 6402 FULL_GRAD
> #Parameter# /_1 {50} 1602 FULL_GRAD
> #Parameter# /_2 {50,50} 80002 FULL_GRAD
> #Parameter# /_3 {50} 1602 FULL_GRAD
> #Parameter# /_4 {100,100} 320002 FULL_GRAD
> #Parameter# /_5 {100} 3202 FULL_GRAD
> #Parameter# /_6 {50,100} 160002 FULL_GRAD
> #Parameter# /_7 {50} 1602 FULL_GRAD
> #Parameter# /_8 {200,200} 1280002 FULL_GRAD
> #Parameter# /_9 {200} 6402 FULL_GRAD
> #Parameter# /_10 {50,200} 320002 FULL_GRAD
> #Parameter# /_11 {50} 1602 FULL_GRAD
> #Parameter# /_12 {300,300} 2880002 FULL_GRAD
> #Parameter# /_13 {300} 9602 FULL_GRAD
> #Parameter# /_14 {50,300} 480002 FULL_GRAD
> #Parameter# /_15 {50} 1602 FULL_GRAD
> #Parameter# /_16 {400,400} 5120002 FULL_GRAD
> #Parameter# /_17 {400} 12802 FULL_GRAD
> #Parameter# /_18 {50,400} 640002 FULL_GRAD
> #Parameter# /_19 {50} 1602 FULL_GRAD
> #Parameter# /_20 {50} 1602 FULL_GRAD
> #LookupParameter# /_0 {100,25541} 81731202 FULL_GRAD
> #LookupParameter# /_21 {50,31013} 49620802 FULL_GRAD

Any help is appreciated.

**Comment by neubig** [comment]
This sort of error normally happens when you have a different model defined at training and test time. I'd make sure that you're calling exactly the same constructor code during training and test. If it still doesn't work, could you provide a runnable code sample so we can debug?

**Comment by peterparkero** [comment]
model.py

    # -*- coding: UTF-8 -*-
    import random,time,os
    from collections import Counter, namedtuple

    import numpy as np
    import dynet as dy

    from tools import initCemb, prepareData
    from test import test

    np.random.seed(970)

    Sentence = namedtuple('Sentence',['score','score_expr','LSTMState','y','prevState','wlen','golden'])

    class CWS (object):
        def __init__(self,Cemb,character_idx_map,options):
            model = dy.Model()
            self.trainer = dy.MomentumSGDTrainer(model,options['lr'],options['momentum']) # we use Momentum SGD
            self.params = self.initParams(model,Cemb,options)
            self.options = options
            self.model = model
            self.character_idx_map = character_idx_map
            self.known_words = None
        
        def load(self,filename):
            self.model.populate(filename)

        def save(self,filename):
            self.model.save(filename)

        def use_word_embed(self,known_words):
            self.known_words = known_words
            self.params['word_embed'] = self.model.add_lookup_parameters((len(known_words),self.options['word_dims']))

        def initParams(self,model,Cemb,options):
            # initialize the model parameters  
            params = dict()
            params['embed'] = model.add_lookup_parameters(Cemb.shape)
            for row_num,vec in enumerate(Cemb):
                params['embed'].init_row(row_num, vec)
            params['lstm'] = dy.LSTMBuilder(1,options['word_dims'],options['nhiddens'],model)
            
            params['reset_gate_W'] = []
            params['reset_gate_b'] = []
            params['com_W'] = []
            params['com_b'] = []

            params['word_score_U'] = model.add_parameters(options['word_dims'])
            params['predict_W'] = model.add_parameters((options['word_dims'],options['nhiddens']))
            params['predict_b'] = model.add_parameters(options['word_dims'])
            for wlen in range(1,options['max_word_len']+1):
                params['reset_gate_W'].append(model.add_parameters((wlen*options['char_dims'],wlen*options['char_dims'])))
                params['reset_gate_b'].append(model.add_parameters(wlen*options['char_dims']))
                params['com_W'].append(model.add_parameters((options['word_dims'],wlen*options['char_dims'])))
                params['com_b'].append(model.add_parameters(options['word_dims']))
            params['<BoS>'] = model.add_parameters(options['word_dims'])
            return params
        
        def renew_cg(self):
            # renew the compute graph for every single instance
            dy.renew_cg()

            param_exprs = dict()
            param_exprs['U'] = dy.parameter(self.params['word_score_U'])
            param_exprs['pW'] = dy.parameter(self.params['predict_W'])
            param_exprs['pb'] = dy.parameter(self.params['predict_b'])
            param_exprs['<bos>'] = dy.parameter(self.params['<BoS>'])

            self.param_exprs = param_exprs
        
        def word_repr(self, char_seq, cembs):
            # obtain the word representation when given its character sequence

            wlen = len(char_seq)
            if 'rgW%d'%wlen not in self.param_exprs:
                self.param_exprs['rgW%d'%wlen] = dy.parameter(self.params['reset_gate_W'][wlen-1])
                self.param_exprs['rgb%d'%wlen] = dy.parameter(self.params['reset_gate_b'][wlen-1])
                self.param_exprs['cW%d'%wlen] = dy.parameter(self.params['com_W'][wlen-1])
                self.param_exprs['cb%d'%wlen] = dy.parameter(self.params['com_b'][wlen-1])

            chars = dy.concatenate(cembs)
            reset_gate = dy.logistic(self.param_exprs['rgW%d'%wlen] * chars + self.param_exprs['rgb%d'%wlen])
            word = dy.tanh(self.param_exprs['cW%d'%wlen] * dy.cmult(reset_gate,chars) + self.param_exprs['cb%d'%wlen])
            if self.known_words is not None and tuple(char_seq) in self.known_words:
                return (word + dy.lookup(self.params['word_embed'],self.known_words[tuple(char_seq)]))/2.
            return word

        def greedy_search(self, char_seq, truth = None, mu =0.):
            init_state = self.params['lstm'].initial_state().add_input(self.param_exprs['<bos>'])
            init_y = dy.tanh(self.param_exprs['pW'] * init_state.output() + self.param_exprs['pb'])
            init_score = dy.scalarInput(0.)
            init_sentence = Sentence(score=init_score.scalar_value(),score_expr=init_score,LSTMState =init_state, y= init_y , prevState = None, wlen=None, golden=True)
            
            if truth is not None:
                cembs = [ dy.dropout(dy.lookup(self.params['embed'],char),self.options['dropout_rate']) for char in char_seq ]
            else:
                cembs = [dy.lookup(self.params['embed'],char) for char in char_seq ]

            start_agenda = init_sentence
            agenda = [start_agenda]

            for idx, _ in enumerate(char_seq,1): # from left to right, character by character
                now = None
                for wlen in range(1,min(idx,self.options['max_word_len'])+1): # generate word candidate vectors
                    # join segmentation sent + word
                    word = self.word_repr(char_seq[idx-wlen:idx], cembs[idx-wlen:idx])
                    sent = agenda[idx-wlen]

                    if truth is not None:
                        word = dy.dropout(word,self.options['dropout_rate'])
                    
                    word_score = dy.dot_product(word,self.param_exprs['U'])

                    if truth is not None:
                        golden =  sent.golden and truth[idx-1]==wlen
                        margin = dy.scalarInput(mu*wlen if truth[idx-1]!=wlen else 0.)
                        score = margin + sent.score_expr + dy.dot_product(sent.y, word) + word_score
                    else:
                        golden = False
                        score = sent.score_expr + dy.dot_product(sent.y, word) + word_score


                    good = (now is None or now.score < score.scalar_value())
                    if golden or good:
                        new_state = sent.LSTMState.add_input(word)
                        new_y = dy.tanh(self.param_exprs['pW'] * new_state.output() + self.param_exprs['pb'])
                        new_sent = Sentence(score=score.scalar_value(),score_expr=score,LSTMState=new_state,y=new_y, prevState=sent, wlen=wlen, golden=golden)
                        if good:
                            now = new_sent
                        if golden:
                            golden_sent = new_sent

                agenda.append(now)
                if truth is not None and truth[idx-1]>0 and (not now.golden):
                    return (now.score_expr - golden_sent.score_expr)

            if truth is not None:
                return (now.score_expr - golden_sent.score_expr)

            return agenda

        def forward(self, char_seq):
            self.renew_cg()
            agenda = self.greedy_search(char_seq)
            now = agenda[-1]
            ans = []
            while now.prevState is not None:
                ans.append(now.wlen)
                now = now.prevState
            return reversed(ans)


        def backward(self, char_seq, truth):
            self.renew_cg()
            loss = self.greedy_search(char_seq,truth,self.options['margin_loss_discount'])
            res = loss.scalar_value()
            loss.backward()
            return res

    def dy_train_model(
        max_epochs = 30,
        batch_size = 256,
        char_dims = 50,
        word_dims = 100,
        nhiddens = 50,
        dropout_rate = 0.2,
        margin_loss_discount = 0.2,
        max_word_len = 4,
        load_params = None,
        max_sent_len = 60,
        shuffle_data = True,
        train_file = '../data/train',
        dev_file = '../data/dev',
        lr = 0.5,
        edecay = 0.1,
        momentum = 0.5,
        pre_trained = '../w2v/char_vecs_100',
        word_proportion = 0.5
    ):
        options = locals().copy()
        print ('Model options:')
        for kk,vv in options.items():
            print ('\t',kk,'\t',vv)

        Cemb, character_idx_map = initCemb(char_dims,train_file,pre_trained)

        cws = CWS(Cemb,character_idx_map,options)

        if load_params is not None:
            cws.load(load_params)
            test(cws, dev_file, 'result')

        char_seq, _ , truth = prepareData(character_idx_map,train_file)
        
        if max_sent_len is not None:
            survived = []
            for idx,seq in enumerate(char_seq):
                if len(seq)<=max_sent_len and len(seq)>1:
                    survived.append(idx)
            char_seq =  [ char_seq[idx]  for idx in survived]
            truth = [ truth[idx] for idx in survived]
        
        if word_proportion > 0:
            word_counter = Counter()
            for chars,labels in zip(char_seq,truth):
                word_counter.update(tuple(chars[idx-label:idx]) for idx,label in enumerate(labels,1))
            known_word_count  = int(word_proportion*len(word_counter))
            known_words =  dict(word_counter.most_common()[:known_word_count])
            idx = 0
            for word in known_words:
                known_words[word] = idx
                idx+=1
            cws.use_word_embed(known_words)

        n = len(char_seq)
        print ('Total number of training instances:',n)
        
        print ('Start training model')
        start_time = time.time()
        nsamples = 0
        for eidx in range(max_epochs):
            idx_list = [i for i in range(n)]
            if shuffle_data:
                np.random.shuffle(idx_list)

            for idx in idx_list:
                loss = cws.backward(char_seq[idx],truth[idx])
                if np.isnan(loss):
                    print ('somthing went wrong, loss is nan.')
                    return
                nsamples += 1
                if nsamples % batch_size == 0:
                    cws.trainer.update()

            cws.trainer.update_epoch(1.)
            end_time = time.time()
            print ('Trained %s epoch(s) (%d samples) took %.lfs per epoch'%(eidx+1,nsamples,(end_time-start_time)/(eidx+1)))
            test(cws,dev_file,'../result/dev_result%d'%(eidx+1))
            os.system('python score.py %s %d %d'%(dev_file,eidx+1,eidx+1))
            cws.save('epoch%d'%(eidx+1))
            print ('Current model saved')


tools.py

    # -*- coding: UTF-8 -*-
    from collections import defaultdict

    import numpy as np
    import random
    import gensim
    import re

    def initCemb(ndims,train_file,pre_trained,thr = 5.):
        f = open(train_file)
        train_vocab = defaultdict(float)
        for line in f.readlines():
            sent = line.split()
            for word in sent:
                for character in word:
                    train_vocab[character]+=1
        f.close()
        character_vecs = {}
        for character in train_vocab:
            if train_vocab[character]< thr:
                continue
            character_vecs[character] = np.random.uniform(-0.5/ndims,0.5/ndims,ndims)
        if pre_trained is not None:
            pre_trained = gensim.models.Word2Vec.load(pre_trained)
            pre_trained_vocab = set([ w for w in pre_trained.wv.vocab.keys()])
            for character in pre_trained_vocab:
                character_vecs[character] = pre_trained[character]
        Cemb = np.zeros(shape=(len(character_vecs)+1,ndims))
        idx = 1
        character_idx_map = dict()
        for character in character_vecs:
            Cemb[idx] = character_vecs[character]
            character_idx_map[character] = idx
            idx+=1
        return Cemb,character_idx_map

    def SMEB(lens):
        idxs = []
        for len in lens:
            for i in range(len-1):
                idxs.append(0)
            idxs.append(len)
        return idxs

    def prepareData(character_idx_map,path,test=False):
        seqs,wlenss,idxss = [],[],[]
        f = open(path)
        for line in f.readlines():
            sent = line.split()
            Left = 0
            for idx,word in enumerate(sent):
                if len(re.sub('\W','',word,flags=re.U))==0:
                    if idx >Left:
                        seqs.append(list(''.join(sent[Left:idx])))
                        wlenss.append([len(word) for word in sent[Left:idx]])
                    Left = idx+1
            if Left!=len(sent):
                seqs.append(list(''.join(sent[Left:])))
                wlenss.append([ len(word) for word in sent[Left:]])
        seqs = [[ character_idx_map[character] if character in character_idx_map else 0 for character in seq] for seq in seqs]
        f.close()
        if test:
            return seqs
        for wlens in wlenss:
            idxss.append(SMEB(wlens))
        return seqs,wlenss,idxss

test.py

    # -*- coding: UTF-8 -*-
    import re
    from tools import prepareData

    def test(cws,filename,output_path):

        def seg(char_seq,text):
            lens = cws.forward(char_seq)
            res, begin =[], 0
            for wlen in lens:
                res.append(''.join(text[begin:begin+wlen]))
                begin+=wlen
            return res

        char_seqs = prepareData(cws.character_idx_map,filename,test=True)
        fo = open(output_path,'wt')
        seq_idx = 0 
        for line in open(filename).readlines():
            sent = line.split()
            Left = 0
            output_sent = []
            for idx,word in enumerate(sent):
                if len(re.sub('\W','',word,flags=re.U))==0:
                    if idx>Left:
                        words =seg(char_seqs[seq_idx],list(''.join(sent[Left:idx])))
                        seq_idx += 1
                        output_sent.extend(words)
                    Left = idx+1
                    output_sent.append(word)
            if Left!=len(sent):
                words = seg(char_seqs[seq_idx],list(''.join(sent[Left:])))
                seq_idx += 1
                output_sent.extend(words)
            output_sent = '  '.join(output_sent)+'\r\n'
            fo.write(output_sent)
        fo.close()

Upon usage of the above programme, I called the following.  Note that I have previously trained a character embedding using gensim (dimension of 100) and saved it into char_embedding.bin.

```
model.dy_train_model(
            max_epochs = 1,
            batch_size = 256,
            char_dims = 100,
            word_dims = 50,
            nhiddens = 50,
            dropout_rate = 0.2,
            max_word_len = 4,
            load_params = None, # None for train mode, otherwise please specify the parameter file.   
            margin_loss_discount = 0.2,
            max_sent_len = 60,
            shuffle_data = True,
            train_file = '../data/cityu_train_all',
            dev_file = '../data/cityu_test',   # dev/test in train/test mode.
            pre_trained = 'char_embedding.bin',
            lr = 0.2,
            edecay = 0.2, #msr,pku 0.2,0.1
            momentum = 0.5,           
            word_proportion= 0.5       
            )
```

The output of the above is as follows:
> Model options:
> 	 word_proportion 	 0.5
> 	 pre_trained 	 char_embedding.bin
> 	 momentum 	 0.5
> 	 edecay 	 0.2
> 	 lr 	 0.2
> 	 dev_file 	 ../data/cityu_test
> 	 train_file 	 ../data/cityu_train_all
> 	 shuffle_data 	 True
> 	 max_sent_len 	 60
> 	 load_params 	 None
> 	 max_word_len 	 4
> 	 margin_loss_discount 	 0.2
> 	 dropout_rate 	 0.2
> 	 nhiddens 	 50
> 	 word_dims 	 50
> 	 char_dims 	 100
> 	 batch_size 	 256
> 	 max_epochs 	 30
> Total number of training instances: 205403
> Start training model
> Trained 1 epoch(s) (205403 samples) took 1189s per epoch
> Current model saved
> Trained 2 epoch(s) (410806 samples) took 1190s per epoch
> Current model saved
> Trained 3 epoch(s) (616209 samples) took 1274s per epoch
> Current model saved
> Trained 4 epoch(s) (821612 samples) took 1286s per epoch
> Current model saved
> Trained 5 epoch(s) (1027015 samples) took 1323s per epoch
> Current model saved
> Trained 6 epoch(s) (1232418 samples) took 1354s per epoch
> Current model saved
> Trained 7 epoch(s) (1437821 samples) took 1348s per epoch
> Current model saved
> Trained 8 epoch(s) (1643224 samples) took 1347s per epoch
> Current model saved
> Trained 9 epoch(s) (1848627 samples) took 1351s per epoch
> Current model saved
> Trained 10 epoch(s) (2054030 samples) took 1345s per epoch
> Current model saved
> Trained 11 epoch(s) (2259433 samples) took 1317s per epoch
> Current model saved
> Trained 12 epoch(s) (2464836 samples) took 1278s per epoch
> Current model saved
> Trained 13 epoch(s) (2670239 samples) took 1240s per epoch
> Current model saved
> Trained 14 epoch(s) (2875642 samples) took 1205s per epoch
> Current model saved
> Trained 15 epoch(s) (3081045 samples) took 1173s per epoch
> Current model saved
> Trained 16 epoch(s) (3286448 samples) took 1145s per epoch
> Current model saved
> Trained 17 epoch(s) (3491851 samples) took 1117s per epoch
> Current model saved
> Trained 18 epoch(s) (3697254 samples) took 1092s per epoch
> Current model saved
> Trained 19 epoch(s) (3902657 samples) took 1071s per epoch
> Current model saved
> Trained 20 epoch(s) (4108060 samples) took 1049s per epoch
> Current model saved
> Trained 21 epoch(s) (4313463 samples) took 1029s per epoch
> Current model saved
> Trained 22 epoch(s) (4518866 samples) took 1013s per epoch
> Current model saved
> Trained 23 epoch(s) (4724269 samples) took 997s per epoch
> Current model saved
> Trained 24 epoch(s) (4929672 samples) took 983s per epoch
> Current model saved
> Trained 25 epoch(s) (5135075 samples) took 971s per epoch
> Current model saved
> Trained 26 epoch(s) (5340478 samples) took 959s per epoch
> Current model saved
> Trained 27 epoch(s) (5545881 samples) took 948s per epoch
> Current model saved
> Trained 28 epoch(s) (5751284 samples) took 937s per epoch
> Current model saved
> Trained 29 epoch(s) (5956687 samples) took 927s per epoch
> Current model saved
> Trained 30 epoch(s) (6162090 samples) took 918s per epoch
> Current model saved

While performing the prediction, I entered the below:

```
model.dy_train_model(
                    max_epochs = 30,
                    batch_size = 256,
                    char_dims = 100,
                    word_dims = 50,
                    nhiddens = 50,
                    dropout_rate = 0.2,
                    max_word_len = 4,
                    load_params = './epoch30', # None for train mode, otherwise please specify the parameter file.   
                    margin_loss_discount = 0.2,
                    max_sent_len = 60,
                    shuffle_data = True,
                    train_file = '../data/cityu_train_all',
                    dev_file = 'char_embedding.bin',   # dev/test in train/test mode.
                    pre_trained = 'char_embedding.bin',
                    lr = 0.2,
                    edecay = 0.2, #msr,pku 0.2,0.1
                    momentum = 0.5,           
                    word_proportion= 0.5       
                    )
```

The difference between the training of the model and performing predictions is on the parameters 'load_params' and 'dev_file', where load_params is changed from None to the file containing the model, and dev_file is changed from '../data/cityu_test' to 'char_embedding.bin'.

> Model options:
> 	 word_proportion 	 0.5
> 	 pre_trained 	 char_embedding.bin
> 	 momentum 	 0.5
> 	 edecay 	 0.2
> 	 lr 	 0.2
> 	 dev_file 	 char_embedding.bin
> 	 train_file 	 ../data/cityu_train_all
> 	 shuffle_data 	 True
> 	 max_sent_len 	 60
> 	 load_params 	 ./epoch30
> 	 max_word_len 	 4
> 	 margin_loss_discount 	 0.2
> 	 dropout_rate 	 0.2
> 	 nhiddens 	 50
> 	 word_dims 	 50
> 	 char_dims 	 100
> 	 batch_size 	 256
> 	 max_epochs 	 30
> AttributeError                            Traceback (most recent call last)
> <ipython-input-15-09dd4d74ebf9> in <module>()
>      17                     edecay = 0.2, #msr,pku 0.2,0.1
>      18                     momentum = 0.5,
> ---> 19                     word_proportion= 0.5
>      20                     )
> 
> ~/greedyCWS-master/src/model.py in dy_train_model(max_epochs, batch_size, char_dims, word_dims, nhiddens, dropout_rate, margin_loss_discount, max_word_len, load_params, max_sent_len, shuffle_data, train_file, dev_file, lr, edecay, momentum, pre_trained, word_proportion)
>     188 
>     189     if load_params is not None:
> --> 190         cws.load(load_params)
>     191         test(cws, dev_file, 'result')
>     192 
> 
> ~/greedyCWS-master/src/model.py in load(self, filename)
>      24 
>      25     def load(self,filename):
> ---> 26         self.model.load(filename)
>      27 
>      28     def save(self,filename):
> 
> AttributeError: '_dynet.ParameterCollection' object has no attribute 'load'

Thanks a lot for the help in advance.

**Comment by neubig** [comment]
Are you sure `character_idx_map` and all other arguments to the constructor CWS are the same during initial training and when you re-load the parameters?

**Comment by neubig** [comment]
Closing for lack of activity.

**Comment by FengZiYjun** [comment]
Hi. This is very likely to be a bug in the author's code, because some additional parameters are added into the model in `cws.use_word_embed` after `cws.load`. I had fixed it in my [fork](https://github.com/FengZiYjun/greedyCWS).
